modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/session/SessionManager.java

```diff
@@ -17,21 +17,44 @@
 
 package org.apache.ignite.internal.sql.engine.session;
 
-import java.util.Map;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
 import java.util.UUID;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.TimeUnit;
 import org.apache.ignite.internal.sql.engine.CurrentTimeProvider;
 import org.apache.ignite.internal.sql.engine.property.PropertiesHolder;
+import org.checkerframework.checker.index.qual.NonNegative;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * A manager of a server side sql sessions.
  */
 public class SessionManager {
-    private final Map<SessionId, Session> activeSessions = new ConcurrentHashMap<>();
     private final CurrentTimeProvider timeProvider;
 
+    /** Active sessions with expiration. */
+    private Cache<SessionId, Session> activeSessions = Caffeine.newBuilder()
+            .weakKeys()
+            .weakValues()
```

Comment:
Does it mean keys/values are reachable if there is at least one non-weak reference on these objects?